torch/_dynamo/convert_frame.py

```diff
@@ -286,6 +285,71 @@ def exception_handler(e, code, frame=None, export=False):
 FRAME_COMPILE_COUNTER: typing.Counter[int] = collections.Counter()
 
 
+def maybe_cprofile(func):
+    if config.cprofile:
+        return cprofile_wrapper(func)
+    return func
+
+
+def cprofile_wrapper(func):
+    @wraps(func)
+    def profile_wrapper(*args, **kwargs):
+        global FRAME_COUNTER
+        global FRAME_COMPILE_COUNTER
+        profile_path = Path(
+            f"/tmp/{func.__name__}_{FRAME_COUNTER}_{FRAME_COMPILE_COUNTER}.profile"
+        )
+        prof = cProfile.Profile()
+        prof.enable()
+        start_ts = time.time()
+        retval = prof.runcall(func, *args, **kwargs)
+        profile_latency = time.time() - start_ts
+        prof.disable()
+        log.info(
+            "### Cprofile for %s iter %d took %.3f seconds ###",
+            func.__name__,
+            profile_cnt,
+            profile_latency,
+        )
+        ps = pstats.Stats(prof)
+        try:
+            prof.dump_stats(profile_path)
+        except PermissionError:
+            log.info("Cannot write to %s", str(profile_path))
+        svg_path = profile_path.with_suffix(".svg")
+        try:
+            gprof2dot_process = subprocess.Popen(
+                [
+                    "gprof2dot",
+                    "-f",
+                    "pstats",
+                    "--node-label=total-time-percentage",
+                    "--node-label=self-time-percentage",
+                    "--node-label=total-time",
+                    str(profile_path),
+                ],
+                stdout=subprocess.PIPE,
+            )
+            subprocess.check_call(
+                ["dot", "-Tsvg", "-o", str(svg_path)],
+                stdin=gprof2dot_process.stdout,
+            )
+            log.info("Generated SVG from profile at %s", str(svg_path))
+        except FileNotFoundError:
+            log.info(
+                "Failed to generate SVG from profile -- dumping stats instead."
+                "Try installing gprof2dot and dot for a better visualization"
+            )
+            ps.sort_stats(pstats.SortKey.TIME).print_stats(20)
+            ps.sort_stats(pstats.SortKey.CUMULATIVE).print_stats(20)
+
+        maybe_upload_prof_stats_to_manifold(str(profile_path))  # fb-only
```

Comment:
When you upload the stats to manifold, you will want to call `trace_structured` to emit the manifold link to the structured logs, and then we can show them in tlparse.

Probably something like:

```
trace_structured(
    "external_url",
    lambda: { "name": "dynamo_cprofile_graphviz", "url": "MANIFOLD_URL" },
)
```

Another option is to directly post the raw profile (if it is plaintext) to the structured log, bypassing manifold (since logs are also just going to manifold).  This would look like:

```
trace_structured(
    "artifact",
    lambda: { "name": "dynamo_cprofile_prof", "type": "prof", "encoding": "base64" },
    payload_fn=lambda: base64.encodebytes(open(profile_file, 'rb').read()),
)
```

and then we would do the gprof2dot/dot postprocessing after the fact. The main benefit to this is if you wanted to change around how you visualized the data, e.g., sending it to snakeviz or something.

Neither of these trace types currently have support in tlparse, but if you add them, we can teach tlparse how to deal with it.

Reply:
thanks @ezyang I will directly post the raw profile to `trace_structured`.  can you show me steps how to test it in internal devserver?

Reply:
Manual testing looks like TORCH_TRACE=/tmp/trace and then manually inspect the log. To close the loop, you should codev the tlparse change too. It's not too hard either; want to learn some Rust?

Reply:
It hits error `Object of type bytes is not JSON serializable`.

[N5332351](https://www.internalfb.com/intern/anp/view/?id=5332351) is a small repro.

Reply:
oh, gotta .encode('ascii') the result of encodebytes lol

Reply:
@ezyang we can start writing Rust of tlparse changes.  can you give some instructions?